crates/network/src/any/mod.rs

```diff
@@ -46,11 +59,515 @@ impl From<TxType> for AnyTxType {
     }
 }
 
+/// Memoization for deserialization of [`AnyTxEnvelope`] and [`AnyTypedTransaction`].
+#[derive(Debug, Clone, PartialEq, Eq, Default)]
+#[allow(unnameable_types)]
+pub struct DeserMemo {
+    input: OnceLock<Bytes>,
+    access_list: OnceLock<AccessList>,
+    blob_versioned_hashes: OnceLock<Vec<B256>>,
+    authorization_list: OnceLock<Vec<SignedAuthorization>>,
+}
+
+/// Transaction envelope for a catch-all network.
+#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
+#[serde(untagged)]
+#[doc(alias = "AnyTransactionEnvelope")]
+pub enum AnyTxEnvelope {
+    /// An Ethereum transaction.
+    Ethereum(TxEnvelope),
+    /// A transaction with unknown type.
+    Other {
+        /// Transaction type.
+        #[serde(rename = "type")]
+        ty: AnyTxType,
+        /// Additional fields.
+        fields: std::collections::BTreeMap<String, serde_json::Value>,
+
+        /// Memoization for deserialization.
+        #[serde(skip, default)]
+        memo: DeserMemo,
+    },
+}
+
+impl Encodable2718 for AnyTxEnvelope {
+    fn type_flag(&self) -> Option<u8> {
+        match self {
+            Self::Ethereum(t) => t.type_flag(),
+            Self::Other { ty, .. } => Some(ty.into()),
+        }
+    }
+
+    fn encode_2718_len(&self) -> usize {
+        match self {
+            Self::Ethereum(t) => t.encode_2718_len(),
+            Self::Other { .. } => 1,
+        }
+    }
+
+    fn encode_2718(&self, out: &mut dyn alloy_primitives::bytes::BufMut) {
+        match self {
+            Self::Ethereum(t) => t.encode_2718(out),
+            Self::Other { ty, .. } => {
+                out.put_u8(ty.into());
```

Comment:
I think a panic might be more appropriate here. this shouldn't be something users would hit accidentally, and a panic would give more context than a mismatched tx hash or rlp encoding

wdyt @mattsse @prestwich

Reply:
i'm concerned that this would turn common user actions like `Provider::send_raw_transaction` into panics. Users are not generally very aware of how transaction construction and serialization works (by design), so unless we can head off calling `send_raw_transaction` in the `AnyNetwork` entirely, I'm wary of making this panic

Reply:
retrieve -> modify -> rebroadcast is a reasonably common user behavior. this could also happen for e.g. searcher bundle, where the txns must be 2718 encoded before being added to a bundle. It is pretty reasonable to expect a searcher to attempt to retrieve a tx from RPC and then re-encode that tx. I'm not sure if it's easier to diagnose as a panic or as an RPC error at that point tho

Reply:
> this could also happen for e.g. searcher bundle, where the txns must be 2718 encoded before being added to a bundle

I'd expect such users to understand the rlp semantics and use a safe network implementation. also not sure whether there are any networks with both custom transaction types and bundle submision support?

> I'm not sure if it's easier to diagnose as a panic or as an RPC error at that point tho

I would say that debugging panic would be much easier than a decoding error from RPC. also imo both scenarios are showstoppers for such projects, so it feels more correct to detect on our side

Reply:
i don't have a strong opinion here. should i change this to a panic with an informative message?

Reply:
we could chose to not do anything which would definitely lead to issues somewhere else
but I think a panic would be more helpful for now and we can consider not writing anything at all if users somehow actually reach this

Reply:
this now panics 👍